Fix ToDate test failure & inefficiency - fixes #3026

tfmorris opened the pull request:
Fixes #3026. 

- Use `Instant.truncate(DAYS)` instead of homebrew offset calculation.
- Fix handling of unknown language code as first format string. It was creating 160 copies of it and trying to convert (unsuccessfully) with them all.

The unknown language code used to return an `EvalError` but that behavior was changed and the test that covered it was disabled. Presumably this was intentional, but I'm not sure why. 
@ostephens do you know why?

tfmorris commented:
I found my code from 2013 https://github.com/OpenRefine/OpenRefine/commit/fa072df85c27ce601f867a7f39c5a5d1bf288dec#diff-c7ab88e74f064de9905c27b627771bdd so I think I understand what this is trying to do now - the first format string can optionally be a language tag. I should be able to work out what needs to be done to fix it.

Marking as Draft for now...

tfmorris commented:
A true fix for the second problem is now included along with a small simplifying refactoring.

wetneb commented on main/src/com/google/refine/expr/functions/ToDate.java:
```diff
@@ -168,8 +167,7 @@ private OffsetDateTime parse(String o1, Locale locale, List<String> formats) {
                 return javax.xml.bind.DatatypeConverter.parseDateTime(o1).getTime().toInstant()
-                		.plusSeconds(ZonedDateTime.now().getOffset().getTotalSeconds())
-                		.atOffset(ZoneOffset.of("Z"));
+                        .truncatedTo(ChronoUnit.DAYS).atOffset(ZoneOffset.of("Z"));
```
It is not clear to me that this new version is equivalent, for instance of datetime objects with precision down to the second. You are correct that the motivation for this `.plusSeconds()` was to preserve the date part across timezones but I think this code also handles timestamps with a higher precision than day.

tfmorris replied:
Can you add a (failing) test that demonstrates the problem? I'm not sure I follow what you're trying to say.

My understanding of the intent was to make sure that `toDate("1984")` ends up pinned to midnight, rather than offset by the number of hours that local time is from UTC. The old code apparently doesn't account for summer time / Daylight Savings Time so, on my system, it ends up at 1 hour past midnight rather than at midnight. The new code fixes that. I'm open to other solutions, but I'm trying to resist having to learn all of the complexities of the new Java date/time system.

tfmorris replied:
I've got a better understanding of what the original problem was and have a better fix. The issue occurred when the test date to be parsed was on one side of the Daylight Savings Time boundary and the current date was on the opposite side, resulting in the offset computed using `now()` to be incorrect. I switched the offset calculation to use the parsed date.

Somehow, in the mean time, the default timezone for Java switched itself from "EDT" to "UTC" so I also changed the test to use a fixed, non-UTC timezone to hopefully catch any other UTC dependencies in the future (although it won't catch temporal shifts like DST).

tfmorris commented:
I think this may affect more than just Daylight Savings Time, e.g. `xsd:DateTime`s which include embedded timezone info, so I need to review the test coverage.

Looking at the code which is transforming to UTC, it feels like the definition of an OpenRefine date has shifted at some point and I'm not sure that was intentional. I think it was always, at least implicitly, a local datetime. This, of course, has problems, but migrating to a different definition should be done in a considered manner. It also seems like there was, perhaps, an opportunity missed in updating to the new Java date support to also update parsing code to parse to the new datatypes directly.

tfmorris commented:
Although I've fixed the timezone offset handling to work as intended, I'm not convinced that the original intention is correct.

I've added a few more tests and one of the tests, that a datetime without a timezone qualifier is interpreted as local, fails because of this code that's casting this one particular case to UTC.

tfmorris commented on main/tests/server/src/com/google/refine/expr/functions/strings/ToFromConversionTests.java:
```diff
@@ -167,12 +170,22 @@ public void testToDate() throws CalendarParserException {
       Assert.assertEquals(invoke("toDate", CalendarParser.parseAsOffsetDateTime("2012-03-01")),CalendarParser.parseAsOffsetDateTime("2012-03-01"));
+
+      // FIXME: Date/times without timezone info should be interpreted as local time, not UTC
+//      Assert.assertEquals(invoke("toDate", "2013-06-01T13:12:11"), CalendarParser.parseAsOffsetDateTime("2013-06-01 13:12:11"));
```
@wetneb I think this test should pass, but it fails due to UTC timezone offset code that you added. Can you review and see if you agree?

thadguidry replied:
@tfmorris I don't want to poke my eyes or mouth into this between both of you only to say this bit of history:
I recall Jacky in a chat asking me about formatter expectations (GREL, cell values, etc.) because he was indeed having issues with correct parsing behavior based on what we expected or not.  That probably should have been a signal for me to ask Antonin or someone else (who was still newish to our team), but I think that discussion did eventually happen between him and Antonin in the contributing PRs and original issues for Java Date that he worked on.  I recall Instant and Locale being 2 terms he asked me about concerning expectations and I didn't have enough expertise.

The ISO_INSTANT would be UTC, yes? https://docs.oracle.com/en/java/javase/13/docs/api/java.base/java/time/format/DateTimeFormatter.html#ISO_INSTANT
But the ISO_LOCAL_DATE_TIME would be local, no? https://docs.oracle.com/en/java/javase/13/docs/api/java.base/java/time/format/DateTimeFormatter.html#ISO_LOCAL_DATE_TIME

This is all I recall of the conversations with him, sorry.  Hope it helps!

thadguidry replied:
Speaking of which...which Resolver style do we feel is most useful for regular OpenRefine users? https://docs.oracle.com/en/java/javase/13/docs/api/java.base/java/time/format/ResolverStyle.html#STRICT

wetneb replied:
@tfmorris that proposed test looks obviously sensible to me.

Broadly speaking, my user-side requirements for this would be:
- if I parse a date string and then extract some date part from it, the value I get should be consistent with what was in the original string
- the rendered version of the date should be consistent with the source string.

I have not checked but that might require changes to `datePart` and cell rendering, to make sure these work correctly for local date times.

tfmorris commented:
Thanks both for the feedback. That's useful background. I think fixing everything is too big for this PR, so I'll leave it as is and start a thread on the dev list to figure out what needs to be done with dates in general. It looks like the switch to interpreting things as UTC vs local was made in May 2018, so there are probably compatibility concerns with both restoring the original behavior or leaving it as is.

Since I didn't get an answer on my question about the reasons for the test behavior change, I've restored the previous behavior of treating a bad format string as an error rather than silently ignoring it. I also restored some other previous tests that had been removed, including those for language tags with variants, e.g. zh-CN, although zh-TW and zh-HK which work on my machine and JDK 8 in Travis don't appear to work for later JDKs.
